tinygrad/runtime/ops_clang.py

```diff
@@ -29,31 +35,41 @@ def emulate_ext_calls(fn, uc, address, size, user_data):
 
 class ClangProgram:
   def __init__(self, name:str, prg:str, binary:bool=False):
+    if binary and DEBUG >= 5: print(prg)
+    extension = args['ext'] if not binary else 'bin'
+    bin_path = self.compile(name, prg, binary=binary, extension=extension)
+    if binary and CI and ARM64:
+      prg_lines = prg.split('\n')
+      self.varsize = align(int(prg_lines[0].split(" ")[1]))
+      self.ext_calls = {(i*4+ADDRESS):ins.split(" ")[1:] for i, ins in enumerate(filter(lambda ins: ins[:4] != 'loop', prg_lines[6:-3])) if ins[:2] == 'bl'}
+      with open(bin_path, "rb") as f:
+        self.prg = f.read()
+    else:
+      self.lib = ctypes.CDLL(bin_path)
+      self.fxn = self.lib[name]
+  @compile_cache("clang",toolchain_hash())
+  def compile(self, name:str, prg:str, binary:bool=False, extension:str=""):
     # TODO: is there a way to not write this to disk?
     # A: it seems there isn't https://stackoverflow.com/questions/28053328/ctypes-cdll-load-library-from-memory-rather-than-file
     #    because ctypes.CDLL() calls dlopen (POSIX) or LoadLibrary (Windows) which require a file
-    fn = f"{tempfile.gettempdir()}/clang_{hashlib.md5(prg.encode('utf-8')).hexdigest()}.{args['ext']}"
-    if binary and DEBUG >= 5: print(prg)
-    if not os.path.exists(fn):
-      tmp = f"{fn}.{os.getpid()}.tmp"
-      if not binary:
-        prg = CLANG_PROGRAM_HEADER + prg
-        subprocess.check_output(args=('clang -shared -O2 -Wall -Werror -x c '+args['cflags']+' - -o '+tmp).split(), input=prg.encode('utf-8'))
-        os.rename(tmp, fn)
-      else:
+    if not binary:
+      prg = CLANG_PROGRAM_HEADER + prg
+      with mktemp_clean() as path:
+        subprocess.check_output(args=('clang -shared -O2 -Wall -Werror -x c '+args['cflags']+' - -o '+path).split(), input=prg.encode('utf-8'))
+        with open(path, "rb") as f:
+          return f.read()
+    else:
+      with mktemp_clean() as as_path, mktemp_clean() as final_path:
```

Comment:
Is there a way to pipe this?

Reply:
With clang I tried to write it to `/dev/stdout` but you get `/usr/bin/ld: final link failed: Illegal seek`.

Oddly enough `cat test.c | clang -shared -O2 -Wall -Werror -x c -lm -fPIC --rtlib=compiler-rt - -o /dev/stdout > test.so` seems to work.

Something like `res = subprocess.check_output(['dd', 'if=/dev/stdin', 'of=/dev/stdout'], input=prg)` also works.